contracts/DKVDaggerHashimoto.sol

```diff
@@ -138,41 +152,88 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
 
     function _fnv256(uint256 a, uint256 b) internal pure returns (uint256) {
         unchecked {
-            return (a * 0x0000000000000000000001000000000000000000000000000000000000000163) ^ b;
+            return
+                (a *
+                    0x0000000000000000000001000000000000000000000000000000000000000163) ^
+                b;
         }
     }
 
     /* ON CHAIN VERFICATION */
     function checkDaggerDataWithProof(
-        uint256 idx,
-        bytes32 kvHash,
+        uint256 chunkIdx,
+        PhyAddr memory kvInfo,
         bytes32[] memory proofs,
-        bytes memory maskedData
-    ) public pure returns (bool) {
-        bytes32 dataHash = keccak256(maskedData);
-        bytes32 rootFromProofs = MerkleLib.calculateRootWithProof(dataHash, idx, proofs);
+        bytes memory unmaskedChunkData
+    ) public view returns (bool) {
+        uint256 chunksNumPerKV = 1 << chunkLenBits;
+        uint256 maxChunkIdx = lastKvIdx * chunksNumPerKV - 1;
+
+        if (chunkIdx > maxChunkIdx) {
+            return keccak256(unmaskedChunkData) == EMPTY_CHUNK_HASH;
+        }
+
+        uint256 mchunkSize = chunkSize;
+        uint256 chunkLeafIdx = chunkIdx % chunksNumPerKV;
+        uint256 maxLeafIdx = MerkleLib.getMaxLeafsNum(
+            kvInfo.kvSize,
+            mchunkSize
+        ) - 1;
+        if (chunkLeafIdx > maxLeafIdx) {
+            return keccak256(unmaskedChunkData) == EMPTY_CHUNK_HASH;
+        }
+
+        // we should consider three cases:
+        // 1. the chunk-leaf is not included in the minTree with empty being full with zero
+        // 2. the chunk-leaf is included in the minTree with the chunk not full of data
+        // 3. the chunk-leaf is included in the minTree with the chunk full of data
+        bytes32 dataHash;
+        if ((chunkLeafIdx * mchunkSize) >= kvInfo.kvSize) {
+            dataHash = bytes32(0);
```

Comment:
I guess we can just check if unmaskedChunkData is zeros for this condition.

Reply:
But it will produce a conflict if the user store chunk-data with zero